### PR TITLE
feat(extras.lang): Add kotlin

### PR DESCRIPTION
**Pull request by arthur-mountain**
### Description:
  Adding kotlin support

---

### The plugins was added

1. `mason`
- [ktlint](https://github.com/pinterest/ktlint)
- [kotlin-language-server](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#kotlin_language_server)
- [kotlin-debug-adapter](https://github.com/fwcd/kotlin-debug-adapter)

---

2. `tree-sitter`
- kotlin

---

3. `nvim-lspconfig`
- kotlin-language-server

---

4. `nvim-lint`(optional)
- ktlint

---

5. `conform`(optional)
- ktlint

---

6. `none-ls`(optional)

---

7. `nvim-dap`(optional)
- [kotlin-debug-adapter](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#kotlin)

---
8. mini.comment/nvim-ts-context-commentstring
- I also opened a [pr](https://github.com/JoosepAlviste/nvim-ts-context-commentstring/pull/104) for kotlin support in nvim-ts-context-commentstring cause I found the commentstring is empty

### CODE REVIEW

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/kotlin.lua:8]:
```suggestion
```
I believe this can be removed, as every server that is defined in `nvim-lspconfig.opts.servers` will be automatically installed.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/kotlin.lua:41]:
```suggestion
    "stevearc/conform.nvim",
    optional = true,
```
`dependencies` are already defined for the plugin. `optional = true` in case someone might choose to use the Extra `none-ls`. It would also be good if you could provide a similar configuration for `none-ls` if possible for formatting.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/kotlin.lua:23]:
```suggestion
```
`dependencies` already defined.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/kotlin.lua:49]:
```suggestion
    "mfussenegger/nvim-dap",
    optional = true,
    dependencies = "williamboman/mason.nvim",
```
It should be only enabled if the user has explicitly enabled the DAP Extra.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (Dogacel) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: However it is "ensure_installed", so doesn't it help with verification? IMO it is better to verify 🙂

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: It's just redundant, since it's already defined in `nvim-lspconfig`. Every server that is defined in `nvim-lspconfig` is automatically added to `mason-lspconfig`'s `ensure_installed` that takes care of the LSP servers' installation. For more info take a look [here](https://github.com/LazyVim/LazyVim/blob/97480dc5d2dbb717b45a351e0b04835f138a9094/lua/lazyvim/plugins/lsp/init.lua#L194-L216).

Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: If servers are defined within nvim-lspconfig, and they are automatically added to mason-lspconfig, then I think we can remove it.

However, everyone might need to have this concept (knowing that servers will be added to mason-lspconfig).

Let me refer to how other plugins of lang define language servers and see if we can implement it in a way that is as consistent as possible with other languages.

Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: > Every server that is defined in nvim-lspconfig is automatically added to mason-lspconfig's ensure_installed that takes care o

Thank you very much for the reminder! It really helped me notice this part!!!

Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: @Dogacel @dpetka2001 
Currently, I see that there are two approaches being used. 

Most of them are only defined in nvim-lspconfig, but for instance, markdown and helm are defined in mason + nvim-lspconfig.

From my perspective, since servers confirmed in nvim-lspconfig will be added to mason-lspconfig, I believe we can consider removing kotlin-language-server from ensured_installed.

What are your thoughts on this～～～?

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:41]: Oh, I see.

I'm sorry for the confusion. Could you please clarify how to ensure that dependencies have been added to Mason? 

I want to make sure they are properly integrated and stable.

I've checked [here](https://github.com/LazyVim/LazyVim/blob/97480dc5d2dbb717b45a351e0b04835f138a9094/lua/lazyvim/plugins/formatting.lua), but I'm unsure where this module is being called from.

Of course, I'll add none-ls！！！

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:49]: Oh, I see now. 

Thank you very much for your suggestion. 

Let me go ahead and make the necessary adjustments.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:41]: > optional = true in case someone might choose to use the Extra none-ls. It would also be good if you could provide a similar configuration for none-ls if possible for formatting.

Thank you very much for your suggestion.

Of course, I'll add none-ls！！！

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: You can test it locally with your PR and see if the server will automatically install or not. Usually, `nvim-lspconfig` is where you define the LSP servers to be installed and `mason.nvim` the rest of the tools needed for a language (formatters and linters). 

Not saying this is absolutely how you have to do it. In my personal opinion, it's just redundant code, that's all. And I don't think it will matter in the end if you also leave it in Mason's `ensure_installed`. I just expressed my opinion. Nothing more than that and the one who will make the final decision is definitely the maintainer as I'm just a simple user like you.

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:41]: You can see in the main LazyVim spec for `conform` that mason is already a dependency. This happens [here](https://github.com/LazyVim/LazyVim/blob/97480dc5d2dbb717b45a351e0b04835f138a9094/lua/lazyvim/plugins/formatting.lua#L35-L36). So, there's no need for you to redefine the dependency, since it's already there.

Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:41]: oh thanks!!! 

I'll remove the dependencies based on what we saw in the formatting.lua!

File: lua/lazyvim/plugins/extras/lang/kotlin.lua
Comment (arthur-mountain) [reply lua/lazyvim/plugins/extras/lang/kotlin.lua:8]: > Not saying this is absolutely how you have to do it. In my personal opinion, it's just redundant code, that's all. And I don't think it will matter in the end if you also leave it in Mason's ensure_installed. 
I agree, I'll remove the kotlin-language-server from the `ensure_installed` section in Mason.

I'll also test it in local to make sure.

**Review by dpetka2001 — APPROVED** [review]
Looks good to me from a LazyVim configuration point of view. I don't do Kotlin however, so am unable to test the configuration. It would be really helpful if you yourself and other users (that do Kotilin) if possible could share their experience with this PR in their local configuration.

Again, I'm just a simple user, so the maintainer has the final say.

**Comment by arthur-mountain** [comment]
@dpetka2001 

I truly appreciate your feedback and assistance!

I'm also not an expert in Kotlin; it just so happened that I wrote some projects with Kotlin, which led me to this PR. 

I hope it can be helpful to users of kotlin.

I also hope to receive feedback from users experienced in Kotlin.

Thank you very much for your help.

**Comment by idelice** [comment]
+1

**Comment by folke** [comment]
Thanks!

**Pull request merged**
